src/rust_connection/mod.rs

```diff
@@ -125,22 +121,22 @@ impl<R: Read, W: Write> RustConnection<R, W> {
     /// reason, you need to pass in a `MutexGuard` to be dropped. This function locks the mutex
     /// again and returns a new `MutexGuard`.
     fn read_packet_and_enqueue<'a>(&'a self, mut inner: MutexGuardInner<'a, W>) -> Result<MutexGuardInner<'a, W>, Box<dyn Error>> {
-        if inner.have_reader {
-            // Someone else is reading; wait for them
-            Ok(self.reader_condition.wait(inner).unwrap())
-        } else {
-            // We can read from the socket
-            inner.have_reader = true;
-            drop(inner);
-
-            let packet = read_packet(&mut *self.read.try_lock().unwrap())?;
-
-            inner = self.inner.lock().unwrap();
-            debug_assert!(inner.have_reader);
-            inner.have_reader = false;
-            inner.enqueue_packet(packet);
-            self.reader_condition.notify_all();
-            Ok(inner)
+        match self.read.try_lock() {
+            Err(_) => {
```

Comment:
This means that `Err(TryLockError::Poisoned(_))` results in a deadlock. Could you update this to also match on the error and do something "more obvious" on `Poisoned`? (Simplest option might be to panic, but it's also possible to return the error to the caller; I think either one is fine here and better than deadlocking.)